Filter and sort web console events table

spadgett opened the pull request:
As-you-type filter and sorting controls for events table. The sort works even when fields are combined into the same column at smaller widths. Uses the `pf-simple-sort` directive for sorting as in @ajacobs21e mockups.

TODO:

- [x] Styling
- [x] Update events directive for individual browse pages
- [x] Show message when all events are hidden
- [x] Filter / sort in controller rather than view

<img width="1006" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/13164682/2eb793ac-d685-11e5-9ffe-4ed8f26be000.png">

Fixes #7445

/cc @smarterclayton @jwforres

smarterclayton commented:
I hadn't seen that sorting control yet - is that the patternfly standard
(or is there not one?)

On Thu, Feb 18, 2016 at 9:22 PM, Sam Padgett <notifications@github.com>
wrote:

> Not ready for code review.
>
> As you type filter and sorting controls for events table. The sort works
> even when fields are combined into the same column at smaller widths. Uses
> the pf-simple-sort directive for sorting as in @ajacobs21e
> <https://github.com/ajacobs21e> mockups.
>
> TODO:
>
>    - Styling
>    - Update events directive for individual browse pages
>
> [image: openshift_web_console]
> <https://cloud.githubusercontent.com/assets/1167259/13164682/2eb793ac-d685-11e5-9ffe-4ed8f26be000.png>
>
> Fixes #7445 <https://github.com/openshift/origin/issues/7445>
>
> /cc @smarterclayton <https://github.com/smarterclayton> @jwforres
> <https://github.com/jwforres>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/openshift/origin/pull/7448
> Commit Summary
>
>    - Filter and sort web console events table
>
> File Changes
>
>    - *M* assets/app/scripts/app.js
>    <https://github.com/openshift/origin/pull/7448/files#diff-0> (1)
>    - *M* assets/app/scripts/controllers/events.js
>    <https://github.com/openshift/origin/pull/7448/files#diff-1> (37)
>    - *M* assets/app/views/events.html
>    <https://github.com/openshift/origin/pull/7448/files#diff-2> (16)
>
> Patch Links:
>
>    - https://github.com/openshift/origin/pull/7448.patch
>    - https://github.com/openshift/origin/pull/7448.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7448>.
>

smarterclayton commented:
Also, nice.

On Thu, Feb 18, 2016 at 11:11 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I hadn't seen that sorting control yet - is that the patternfly standard
> (or is there not one?)
>
> On Thu, Feb 18, 2016 at 9:22 PM, Sam Padgett <notifications@github.com>
> wrote:
>
>> Not ready for code review.
>>
>> As you type filter and sorting controls for events table. The sort works
>> even when fields are combined into the same column at smaller widths. Uses
>> the pf-simple-sort directive for sorting as in @ajacobs21e
>> <https://github.com/ajacobs21e> mockups.
>>
>> TODO:
>>
>>    - Styling
>>    - Update events directive for individual browse pages
>>
>> [image: openshift_web_console]
>> <https://cloud.githubusercontent.com/assets/1167259/13164682/2eb793ac-d685-11e5-9ffe-4ed8f26be000.png>
>>
>> Fixes #7445 <https://github.com/openshift/origin/issues/7445>
>>
>> /cc @smarterclayton <https://github.com/smarterclayton> @jwforres
>> <https://github.com/jwforres>
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/openshift/origin/pull/7448
>> Commit Summary
>>
>>    - Filter and sort web console events table
>>
>> File Changes
>>
>>    - *M* assets/app/scripts/app.js
>>    <https://github.com/openshift/origin/pull/7448/files#diff-0> (1)
>>    - *M* assets/app/scripts/controllers/events.js
>>    <https://github.com/openshift/origin/pull/7448/files#diff-1> (37)
>>    - *M* assets/app/views/events.html
>>    <https://github.com/openshift/origin/pull/7448/files#diff-2> (16)
>>
>> Patch Links:
>>
>>    - https://github.com/openshift/origin/pull/7448.patch
>>    - https://github.com/openshift/origin/pull/7448.diff
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/7448>.
>>
>
>

ajacobs21e commented:
@smarterclayton it's a mix of the controls from patternfly's data list
https://www.patternfly.org/patterns/list-view/

with the data table https://www.patternfly.org/patterns/table-view/

Events are probably better suited for a data table, especially since "kind" and "reason" aren't intuitive column names. Although maybe we should reconsider those anyway.

spadgett commented:
@ajacobs21e Ah, I'm using the data list sort on a table. There's no data table directive in angular-patternfly, which makes it difficult for us to use.

How does data table look at mobile sizes? @sg00dwin did a lot of work to change the table layout at different sizes, and I'm not sure how clicking column headers would work with that. For example,

<img width="508" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/13175593/6e2ad4dc-d6d9-11e5-9889-7be679697b9a.png">

spadgett commented:
I also like that we'd have consistent sorting everywhere if we switch to data list for the other browse pages.

ajacobs21e commented:
@spadgett I checked the sample data table here https://rawgit.com/patternfly/patternfly/master/tests/basic.html
and it horizontal scrolls at mobile which I do not like. 

In your example we'd want to keep the sort control. @rhamilto and I have talked about doing an inventory of all the tables in the UI and determining if data list or data table is the best for each. I don't want to say we'll swtich to data list for all without thinking through all the implications. I can do a first pass of this next week while Rob is gone and then the two of us can review when he's back. Will that work?

spadgett commented:
> I can do a first pass of this next week while Rob is gone and then the two of us can review when he's back. Will that work?

Sounds good, we won't be able to use data lists anyway until we upgrade Patternfly.

spadgett commented:
Added to the individual resource pages:

<img width="751" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/13203847/d58f35d0-d88e-11e5-8307-43211040288f.png">

spadgett commented:
I need to add a divider between the filter input and the sort dropdown as in the pf list view example. Otherwise, it looks like the dropdown is for the filter.

spadgett commented:
Added the divider and adjusted the filter width.

<img width="553" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/13204226/8a819fa6-d898-11e5-9366-799326dde410.png">

spadgett commented:
@jwforres PTAL

Torn between trying to refactor the browse events page to reuse the events directive. There's some similar code between them, but they're just different enough it might not be worth it.

smarterclayton commented:
Glorious

On Sun, Feb 21, 2016 at 12:56 PM, Sam Padgett <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> PTAL
>
> Torn between trying to refactor the browse events page to reuse the events
> directive. There's some similar code between them, but they're just
> different enough it might not be worth it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7448#issuecomment-186874145>.
>

jwforres commented on assets/app/scripts/controllers/events.js:
```diff
@@ -8,22 +8,76 @@
+        var watches = [];
+        watches.push(DataService.watch("events", context, function(data) {
+          $scope.unfilteredEvents = data.by("metadata.name");
+          update();
```
Should we debounce the call to update?  Just worried about how often we are going to be re-sorting and re-filtering this list in cases where we are getting spammed with events, especially repeating events where the count just keeps increasing

liggitt replied:
+1

jwforres commented:
@spadgett were you thinking of putting this in for 1.2?

spadgett commented:
No, I was thinking post 1.2.

jwforres commented:
K might be worth refactoring then, will take a look at the difference
tomorrow.
On Feb 21, 2016 8:57 PM, "Sam Padgett" <notifications@github.com> wrote:

> No, I was thinking post 1.2.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7448#issuecomment-186968239>.
>

liggitt commented on assets/app/scripts/directives/events.js:
```diff
@@ -11,16 +11,66 @@ angular.module('openshiftConsole')
+
+        var applyFilter = $filter('filter');
+        var filterForKeyword = function() {
+          $scope.filteredEvents = applyFilter(sortedEvents, $scope.filter.text);
```
throttle here as well? if I type 10 letters in a second, filtering 4 times should be plenty

jwforres commented:
So big differences between events directive and events page:

* Events directive currently expects a resourceKind and resourceName, but easy enough to make that optional and have filterForResource check for their existence
* Additional Name and Reason columns that collapse with other columns in a particular way at mobile, can trigger showing/hiding these depending on whether resourceName and resourceKind are set.  The mobile part will be the most annoying but still doable.
* Sort options depend on which set of columns are being used

@spadgett anything else big?  it seems worthwhile to do it and then we are only maintaining this stuff in one place

spadgett commented:
@jwforres I'll give the refactor a shot and debounce updates.

spadgett commented:
I've done the refactor, not as bad as I thought. I'll push an update shortly.

spadgett commented:
@jwforres updated

spadgett commented:
Rebased.

jwforres commented on assets/app/scripts/directives/events.js:
```diff
@@ -5,22 +5,106 @@ angular.module('openshiftConsole')
         var watches = [];
         watches.push(DataService.watch("events", $scope.projectContext, function(events) {
           $scope.emptyMessage = "No events to show";
           // $scope.eventsArray = $filter('toArray')(events.by("metadata.name"));
```
know why this commented line is here?

spadgett replied:
Yeah, it wasn't needed since `_.filter` handles objects. Comment I made on another PR. I'll remove.

jwforres commented on assets/app/scripts/directives/events.js:
```diff
@@ -5,22 +5,106 @@ angular.module('openshiftConsole')
+        };
+
+        // Conditionally add kind and name to sort fields if not passed to the directive.
+        if (!$scope.resourceKind || !$scope.resourceName) {
```
this wouldn't get updated if the data is set later on the directive right? since we aren't watching it

spadgett replied:
That's correct. The `pf-simple-sort` widget doesn't support changing config currently, although I think it's been fixed in Patternfly 3.1.0.

jwforres commented on assets/app/scripts/filters/resources.js:
```diff
@@ -781,6 +781,12 @@ angular.module('openshiftConsole')
+      }
+
+      // Handle kinds like "ReplicationController" as well
+      resourceType = resourceType.toLowerCase();
```
APIService.kindToResource

jwforres replied:
although we are explicitly humanizing differently here when we want the singular...

spadgett replied:
The map treats "replicationcontroller" and "replicationcontrollers" differently, so that will break existing callers. We need to clean it up when we fix the other kind/resource bug.

spadgett replied:
I can add a `humanizeKind` filter and leave this one alone?

jwforres replied:
yeah... humanizeKind and it should just split on the upper case letters

jwforres commented on assets/app/views/directives/events.html:
```diff
@@ -1,37 +1,80 @@
+        <!-- Only show Name and Kind columns if not passed to the directive. -->
+        <th ng-if="!resourceKind || !resourceName">
+          <span class="hidden-xs-inline visible-sm-inline visible-md-inline hidden-lg-inline">Resource</span>
+          <span class="visible-lg-inline">Name</span>
```
in the old events table didn't we combine Name and Kind at mobile just like we do with Reason and Message

spadgett replied:
We still are. I changed the header to "Resource" from "Kind and Name". I can change it back if that's confusing.

jwforres replied:
sorry not mobile, i think it was tablet res?

jwforres replied:
sorry github didnt refresh didnt see your response.  it won't say "Resource" in the filter though

jwforres replied:
i mean in the sort dropdown

spadgett replied:
No, it is still Kind and Name in the dropdown. I'll change it back.

jwforres commented:
nothing else, but it needs to wait for #7475 since it has css changes

spadgett commented:
@jwforres updated

spadgett commented:
Rebased.

jwforres commented:
Since we are humanizing the kind, someone might type "Replication Controller" into the filter box, does that do what they would expect and show all the RCs?

jwforres commented:
just the one question, otherwise LGTM

spadgett commented:
> Since we are humanizing the kind, someone might type "Replication Controller" into the filter box, does that do what they would expect and show all the RCs?

It won't work unfortunately, but you probably won't get that far since we filter as you type. By the time you type "rep", it's usually filtered down.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/13329160/64cb8396-dbc0-11e5-933c-02e6a4e36dea.png)

We have a similar issue with reasons. I'm using the built-in Angular `filter` filter.

spadgett commented:
Let me see if there's a different expression that will work.

https://code.angularjs.org/1.3.19/docs/api/ng/filter/filter

spadgett commented:
@jwforres I have the filter matching multiple words. Another look?

jwforres commented:
i dont love looping but i cant think of a better way to do that, it will be fine unless they paste some ridiculous string in there like "a a a a a a a a ...."

spadgett commented:
At least the second loop only looks at the results matching the first keyword.

liggitt commented:
dedupe their search tokens and sort by length, matching longest first?

jwforres commented:
do the dedup and sort after they type and store it, so you dont have to
re-do it on event updates

On Thu, Feb 25, 2016 at 4:12 PM, Jordan Liggitt <notifications@github.com>
wrote:

> dedupe their search tokens and sort by length, matching longest first?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7448#issuecomment-188990908>.
>

spadgett commented:
updated

jwforres commented:
k lgtm [merge]

spadgett commented:
[test]

spadgett commented:
[merge]

spadgett commented:
[merge]
